README.md
diana-compose-email is a utility program providing a gui for sending
an email with attachments. It is intended for use with diana.

License: GPL-2


debian/changelog
diana-compose-email (0.9-1) unstable; urgency=low

  * updates for jammy

 -- Alexander Bürger <alexander.buerger@met.no>  Mon, 29 Aug 2022 08:00:08 +0200

diana-compose-email (0.8-3) unstable; urgency=low

  * build-depend update for bionic

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 28 Nov 2018 08:00:08 +0100

diana-compose-email (0.8-2) unstable; urgency=low

  * build-depend on qt5 packages

 -- Alexander Bürger <alexander.buerger@met.no>  Tue, 09 May 2017 08:09:10 +0200

diana-compose-email (0.8-1) unstable; urgency=low

  * use qt5

 -- Alexander Bürger <alexander.buerger@met.no>  Tue, 09 May 2017 08:09:10 +0200

diana-compose-email (0.7-1) unstable; urgency=low

  * translation updates

 -- Alexander Bürger <alexander.buerger@met.no>  Fri, 20 Jan 2017 08:09:10 +0100

diana-compose-email (0.6-1) unstable; urgency=low

  * cmake build problems

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 18 Jan 2017 10:12:14 +0200

diana-compose-email (0.5-1) unstable; urgency=low

  * cmake build problems

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 18 Jan 2017 10:11:12 +0200

diana-compose-email (0.4-1) unstable; urgency=low

  * CMake build, manpage, icon

 -- Alexander Bürger <alexander.buerger@met.no>  Tue, 17 Jan 2017 14:15:16 +0200

diana-compose-email (0.3-1) unstable; urgency=low

  * bugfix translation files

 -- Alexander Bürger <alexander.buerger@met.no>  Mon, 16 Jan 2017 15:45:15 +0200

diana-compose-email (0.2-1) unstable; urgency=low

  * support for dragging of attachments to other programs

 -- Alexander Bürger <alexander.buerger@met.no>  Mon, 16 Jan 2017 15:00:15 +0200

diana-compose-email (0.1-1) unstable; urgency=low

  * extracted from diana

 -- Alexander Bürger <alexander.buerger@met.no>  Tue, 03 Jan 2017 08:00:08 +0200


src/main.cc

#include "compose_email.h"

#include <QApplication>
#include <QLocale>
#include <QProcess>
#include <QTranslator>

int main(int argc, char* argv[])
{
    QApplication app(argc, argv);
    const QString locale = QLocale::system().name();
    QTranslator translator;
    translator.load(QString(":/translations/compose_email_") + locale);
    app.installTranslator(&translator);

    QString tmpdir, from;

    int a = 1;
    for (; a < argc; ++a) {
        QString arg(argv[a]);
        if (arg == "--")
            break;
        if (arg == "-tmpdir") {
            if (a+1 < argc) {
                tmpdir = argv[a+1];
                a += 1;
            }
        }
        if (arg == "-from") {
            if (a+1 < argc) {
                from = argv[a+1];
                a += 1;
            }
        }
    }
    a += 1; // skip "--"

    MailDialog dialog;
    if (!from.isEmpty())
        dialog.setFrom(from);
    for (; a<argc; ++a)
        dialog.attach(argv[a]);
    dialog.exec();

    if (!tmpdir.isEmpty()) {
        QProcess::execute("rm", QStringList() << "-rvf" << tmpdir);
    }

    return 0;
}


src/compose_email.h
/*
  Diana - A Free Meteorological Visualisation Tool

  Copyright (C) 2006-2017 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of Diana

  Diana is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  Diana is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Diana; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/
#ifndef _maildialog_h
#define _maildialog_h

#include <QDialog>
#include <QDialogButtonBox>
#include <QGroupBox>
#include <QLabel>
#include <QLineEdit>
#include <QStringList>
#include <QTextEdit>
#include <QVBoxLayout>

/**
   \brief Dialogue for sending a picture via email
*/
class MailDialog: public QDialog
{ Q_OBJECT;

public:
  MailDialog(QWidget* parent=0);

  void setFrom(const QString& from)
    { from_ = from; }

  void attach(const QString& filename)
    { attachments_ << filename; }

private:
 void createGridGroupBox();

protected:
  void accept(); //Overloaded to perform mail send

private:
  QGroupBox *gridGroupBox;
  QLabel *eToLabel;
  QLabel *eCcLabel;
  QLabel *eSubjectLabel;
  QLineEdit *eToEdit;
  QLineEdit *eCcEdit;
  QLineEdit *eSubjectEdit;
  QTextEdit *eTextEdit;
  QDialogButtonBox *buttonBox;
  QAction *exitAction;

  QString from_;
  QStringList attachments_;
};

#endif


src/compose_email_de.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="de">
<context>
    <name>MailDialog</name>
    <message>
        <location filename="compose_email.cc" line="74"/>
        <source>E-Mail Picture</source>
        <translation>Bild als Email verschicken</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="89"/>
        <source>E-Mail Details</source>
        <translation>E-Mail Angabe-Mail Angaben</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="92"/>
        <source>To:</source>
        <translation>An:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="97"/>
        <source>Cc:</source>
        <translation>Kopie:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="102"/>
        <source>Subject:</source>
        <translation>Betreff:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="109"/>
        <source>Message</source>
        <translation>Nachricht</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="115"/>
        <source>Attachments</source>
        <translation>Anhänge</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Cannot send mail</source>
        <translation>Email-Versand nicht möglich</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Unable to start sendmail program.</source>
        <translation>Kann Programm sendmail nicht starten.</translation>
    </message>
</context>
</TS>


src/compose_email_es.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="es">
<context>
    <name>MailDialog</name>
    <message>
        <location filename="compose_email.cc" line="74"/>
        <source>E-Mail Picture</source>
        <translation>Enviar Figura por correo eletrónico</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="89"/>
        <source>E-Mail Details</source>
        <translation>Enviar detalles por correo electrónico</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="92"/>
        <source>To:</source>
        <translation type="unfinished">A:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="97"/>
        <source>Cc:</source>
        <translation type="unfinished">Copia:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="102"/>
        <source>Subject:</source>
        <translation type="unfinished">Asunto:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="109"/>
        <source>Message</source>
        <translation type="unfinished">Mensaje</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="115"/>
        <source>Attachments</source>
        <translation type="unfinished">Archivos adjuntos</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Cannot send mail</source>
        <translation></translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Unable to start sendmail program.</source>
        <translation type="unfinished">Incapaz de iniciar el programa sendmail.</translation>
    </message>
</context>
</TS>


src/compose_email_nb.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="nb">
<context>
    <name>MailDialog</name>
    <message>
        <location filename="compose_email.cc" line="74"/>
        <source>E-Mail Picture</source>
        <translation>Send bilde som epost</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="89"/>
        <source>E-Mail Details</source>
        <translation>Epostdetaljer</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="92"/>
        <source>To:</source>
        <translation>Til:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="97"/>
        <source>Cc:</source>
        <translation>Kopi:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="102"/>
        <source>Subject:</source>
        <translation>Emne:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="109"/>
        <source>Message</source>
        <translation>Melding</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="115"/>
        <source>Attachments</source>
        <translation>Vedlegg</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Cannot send mail</source>
        <translation>Kan ikke sende epost</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Unable to start sendmail program.</source>
        <translation>Kan ikke starte sendmail programmet.</translation>
    </message>
</context>
</TS>


src/compose_email_ru.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="ru">
<context>
    <name>MailDialog</name>
    <message>
        <location filename="compose_email.cc" line="74"/>
        <source>E-Mail Picture</source>
        <translation>Отправить изображение</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="89"/>
        <source>E-Mail Details</source>
        <translation>Детали отправки</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="92"/>
        <source>To:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="97"/>
        <source>Cc:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="102"/>
        <source>Subject:</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="109"/>
        <source>Message</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="115"/>
        <source>Attachments</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Cannot send mail</source>
        <translation></translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Unable to start sendmail program.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>


src/compose_email_sv.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="sv">
<context>
    <name>MailDialog</name>
    <message>
        <location filename="compose_email.cc" line="74"/>
        <source>E-Mail Picture</source>
        <translation>Skicka bild som epost</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="89"/>
        <source>E-Mail Details</source>
        <translation>Epostdetaljer</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="92"/>
        <source>To:</source>
        <translation>Till:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="97"/>
        <source>Cc:</source>
        <translation>Kopia:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="102"/>
        <source>Subject:</source>
        <translation>Ämne:</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="109"/>
        <source>Message</source>
        <translation>Meddelande</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="115"/>
        <source>Attachments</source>
        <translation>Bilaga</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Cannot send mail</source>
        <translation>Kan inte sända epost</translation>
    </message>
    <message>
        <location filename="compose_email.cc" line="143"/>
        <source>Unable to start sendmail program.</source>
        <translation>Kan inte starta sendmail programmet.</translation>
    </message>
</context>
</TS>


src/compose_email.cc
/*
  Diana - A Free Meteorological Visualisation Tool

  Copyright (C) 2006-2017 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of Diana

  Diana is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  Diana is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Diana; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "compose_email.h"

#include <QByteArray>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QGridLayout>
#include <QImage>
#include <QListView>
#include <QMessageBox>
#include <QTabWidget>
#include <QTemporaryFile>

// #define WRITE_TO_FILE

//******************************************************************************
//* MailDialog::MailDialog( QWidget* parent, Controller* llctrl )
//*  : QDialog(parent), m_ctrl(llctrl)
//*
//*   Purpose : MailDialog constructor
//*   To call :
//*   Returns :
//*      Note :
//******************************************************************************
MailDialog::MailDialog(QWidget* parent)
    : QDialog(parent)
    , from_("diana_noreply@met.no")
{
  //--- Create dialog elements ---
    createGridGroupBox();
    buttonBox = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    //--- Connect buttons to actions ---
    connect(buttonBox, SIGNAL(accepted()), this, SLOT(accept()));
    connect(buttonBox, SIGNAL(rejected()), this, SLOT(reject()));
    //--- Create dialog ---
    QVBoxLayout *mainLayout = new QVBoxLayout;
    mainLayout->addWidget(gridGroupBox);
    mainLayout->addWidget(buttonBox);
    setLayout(mainLayout);
    //--- Set dialog title ---
    setWindowTitle(tr("E-Mail Picture"));
    setWindowIcon(QIcon(":/icons/compose-email.svg"));
}

//******************************************************************************
//* Maildialog::createGridGroupBox()
//*
//*   Purpose : Creates dialog layout
//*   To call : Nothing
//*   Returns : Nothing
//*      Note :
//******************************************************************************
void MailDialog::createGridGroupBox()
{
    //--- Group box containg TO:, CC: & Subject elements ---
    gridGroupBox = new QGroupBox(tr("E-Mail Details"));
    QGridLayout *layout = new QGridLayout;
    //--- Create text & input box for TO: ---
    eToLabel = new QLabel("TO : ");
    eToEdit = new QLineEdit;
    layout->addWidget(eToLabel, 1, 0, Qt::AlignRight);
    layout->addWidget(eToEdit, 1, 1);
    //--- Create text & input box for CC: ---
    eCcLabel = new QLabel("CC : ");
    eCcEdit = new QLineEdit;
    layout->addWidget(eCcLabel, 2, 0, Qt::AlignRight);
    layout->addWidget(eCcEdit, 2, 1);
    //--- Create text & input box for Subject: ---
    eSubjectLabel = new QLabel("Subject : ");
    eSubjectEdit = new QLineEdit;
    layout->addWidget(eSubjectLabel, 3, 0, Qt::AlignRight);
    layout->addWidget(eSubjectEdit, 3, 1);
    //---
    QTabWidget* tabs = new QTabWidget(this);
    eTextEdit = new QTextEdit;
    tabs->addTab(eTextEdit, tr("Message"));
    attachments_ = new AttachedFilesModel(this);
    QListView *listView = new QListView(this);
    listView->setSelectionMode(QAbstractItemView::ExtendedSelection);
    listView->setDragEnabled(true);
    listView->setModel(attachments_);
    tabs->addTab(listView, tr("Attachments"));
    layout->addWidget(tabs, 4, 0, 1, 2);
    //--- Set layout manager for widget ---
    gridGroupBox->setLayout(layout);
}

void MailDialog::attach(const QString& filename)
{
  attachments_->attach(filename);
}

//******************************************************************************
//* Maildialog::accept()
//*
//*   Purpose : Sends mail
//*   To call : Nothing
//*   Returns : Nothing
//*      Note :
//******************************************************************************
void MailDialog::accept()
{
#ifdef WRITE_TO_FILE
    FILE* sendmail = fopen("/tmp/mail.txt", "w");
#else
    FILE* sendmail = popen("/usr/lib/sendmail -t", "w");
#endif

    if (!sendmail) {
      QMessageBox::critical(this, tr("Cannot send mail"), tr("Unable to start sendmail program."));
      return;
    }

    //--- Mail header ---
    fprintf(sendmail,"MIME-Version: 1.0\n");
    fprintf(sendmail,"From: %s\n", from_.toStdString().c_str());
    fprintf(sendmail,"To: %s\n", eToEdit->text().toStdString().c_str());
    fprintf(sendmail,"Cc: %s\n", eCcEdit->text().toStdString().c_str());
    fprintf(sendmail,"Subject: %s\n", eSubjectEdit->text().toStdString().c_str());
    fprintf(sendmail,"Content-Type: multipart/mixed; boundary=\"diana_auto_generated\"\n\n");
    fprintf(sendmail,"This is a multi-part, MIME format, message.\n");
    //--- Mail body ---
    fprintf(sendmail,"--diana_auto_generated\n");
    fprintf(sendmail,"Content-type: text/plain\n\n");
    fprintf(sendmail,"%s\n\n", eTextEdit->toPlainText().toStdString().c_str());

    //--- Mail attachments ---
    const QStringList& filenames = attachments_->attachments();
    Q_FOREACH (const QString& filename, filenames) {
        //--- read the image and convert to Base64 byte array ---
        QFile file(filename);
        if (!file.open(QIODevice::ReadOnly)){
            return;
        }
        QDataStream in(&file);
        QByteArray rawarray;
        QDataStream datastream(&rawarray, QIODevice::WriteOnly);
        char a;
        // copy between the two data streams
        while (in.readRawData(&a,1) != 0){
            datastream.writeRawData(&a,1);
        }
        QByteArray base64array = rawarray.toBase64();

        QFileInfo fi(filename);
        const std::string basename = fi.fileName().toStdString();

        const QString suffix = fi.suffix();
        const char* mime;
        if (suffix == "png")
            mime = "image/png";
        else if (suffix == "gif")
            mime = "image/gif";
        else if (suffix == "pdf")
            mime = "application/pdf";
        else
            mime = "application/octet-stream";

        fprintf(sendmail,"--diana_auto_generated\n");
        fprintf(sendmail,"Content-type: %s; name=\"%s\"\n", mime, basename.c_str());
        fprintf(sendmail,"Content-Transfer-Encoding: base64\n");
        fprintf(sendmail,"Content-Disposition: inline; filename=\"%s\"\n\n", basename.c_str());
        //--- the image data ---
        const char *data = base64array.constData();
        fprintf(sendmail,"%s",data);
        fprintf(sendmail,"\n\n");
    }
    fprintf(sendmail,"--diana_auto_generated--\n\n");

    //--- Finished with sendmail ---
#ifdef WRITE_TO_FILE
    fclose(sendmail);
#else
    pclose(sendmail);
#endif
    QDialog::accept();
}
